Tests/FluentAssertions.Specs/Equivalency/DataEquivalencySpecs.DataColumn.cs

```diff
@@ -0,0 +1,713 @@
+﻿using System;
+using System.Data;
+
+using Xunit;
+using Xunit.Sdk;
+
+namespace FluentAssertions.Specs
+{
+    /// <summary>
+    /// DataColumnEquivalency specs.
+    /// </summary>
+    public partial class DataEquivalencySpecs
+    {
+        public class DataColumnEquivalencySpecs : DataEquivalencySpecs
+        {
+            [Fact]
+            public void When_DataColumns_are_identical_it_should_succeed()
+            {
+                // Arrange
+                var dataSet1 = CreateDummyDataSet<TypedDataSetSubclass>();
+                var dataSet2 = new TypedDataSetSubclass(dataSet1);
+
+                var dataTable1 = dataSet1.TypedDataTable1;
+                var dataTable2 = dataSet2.TypedDataTable1;
+
+                // Act & Assert
+                dataTable1.RowIDColumn.Should().BeEquivalentTo(dataTable2.RowIDColumn);
+            }
+
+            [Fact]
+            public void When_DataColumns_are_both_null_it_should_succeed()
+            {
+                // Act & Assert
+                ((DataColumn)null).Should().BeEquivalentTo(null);
+            }
+
+            [Fact]
+            public void When_DataColumn_is_null_and_isnt_expected_to_be_it_should_fail()
+            {
+                // Arrange
+                var dataSet = CreateDummyDataSet<TypedDataSet>();
+
+                var dataTable = dataSet.TypedDataTable1;
+
+                // Act
+                Action action = () => ((DataColumn)null).Should().BeEquivalentTo(dataTable.RowIDColumn);
+
+                // Assert
+                action.Should().Throw<NullReferenceException>();
```

Comment:
This should throw an `XUnitException`.
See e.g. `When_subject_is_null_it_should_throw` from `BasicEquivalencySpecs.cs`

Reply:
Huh, turns out this `NullReferenceException` is coming from `GenericDictionaryEquivalenceStep.cs`, because it is assuming that `context.CompileTimeType` won't be `null`.

Reply:
Looks like the `null` being passed in by `BasicEquivalencySpecs.cs` gets caught by one of the earlier equivalency steps, and steps that come after don't handle `null` values well. In order for the `DataColumn` equivalency to handle `null`, those intermediate steps need to handle it gracefully as well. I think that's what's going on, anyway.

Reply:
Okay, so, it looks like the _real_ problem is that the top-level `Should` methods aren't actually obtaining a "compile-time" type. Because of the subclassing involved in typed data tables, they're pulling the runtime type. Maybe this can be corrected there using generics.

Reply:
Okay, I've done some work in that direction, and bumped into, once again, the problem of `Should<T>` only being allowed once. So, we can't have a `.Should<TDataTable>` _and_ a `.Should<TDataSet>`. I can't see a way to get the _actual_ compile-time type implicitly in these cases without `Should` having a different name for each type. This is dissatisfying. :-/

Here's the best I can think of: Have it initially just take the base type as the compile-time type, and then have an `As<T>` method on `DataSetAssertions`, `DataTableAssertions` and `DataRowAssertions` that supports setting the compile-time type to a derived type.

Reply:
I have made a commit to this effect, let me know your thoughts.

8cba5f4d26c2719784f756c11bde5761c0a5a4b1